### PR TITLE
Timeline and Milestone REST API + Misc Fixes

### PR DESCRIPTION
**Pull request by ngoctay**
- Merge the code with "topcoder Connect - Timeline and Milestone REST API" challenge (30065695)
- (Review) Milestone: startDate, status should be required fields
- (Review) When deleting a timeline, it should be cascaded to it's milestone as well
- (Review) Milestone completionDate must not be before startDate
- (Review) Postman: DELETE should not have request body
- (Review) Postman: add at least 1 negative test case for the POST end points
- (Paulo) Don’t allow startDate and endDate of milestones out of bound of it’s parent timeline.
  If startDate or endDate for timeline is updated, children must validated and changed accordingly
- (Paulo) GET /timelines, /timelines/{id} should return the associated milestones
- (Paulo) Update milestone order for PATCH endpoint

### CODE REVIEW

**Comment by ngoctay** [comment]
@coderReview @vikasrohit Please have a look.

File: src/models/milestone.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * The Milestone model
+ */
+module.exports = (sequelize, DataTypes) => {
+  const Milestone = sequelize.define('Milestone', {
+    id: { type: DataTypes.BIGINT, primaryKey: true, autoIncrement: true },
+    name: { type: DataTypes.STRING(255), allowNull: false },
+    description: DataTypes.STRING(255),
+    duration: { type: DataTypes.INTEGER, allowNull: false },
+    startDate: { type: DataTypes.DATE, allowNull: false },
```
Comment (ngoctay) [inline src/models/milestone.js]: (Review) Milestone: startDate, status should be required fields

File: src/routes/timelines/delete.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  permissions('timeline.delete'),
+  (req, res, next) => {
+    const timeline = req.timeline;
+    const deleted = _.omit(timeline.toJSON(), ['deletedAt', 'deletedBy']);
+
+    return models.sequelize.transaction(() =>
+      // Update the deletedBy, then delete
+      timeline.update({ deletedBy: req.authUser.userId })
+        .then(() => timeline.destroy())
+        // Cascade delete the milestones
```
Comment (ngoctay) [inline src/routes/timelines/delete.js]: (Review) When deleting a timeline, it should be cascaded to it's milestone as well

File: src/routes/milestones/create.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  },
+  body: {
+    param: Joi.object().keys({
+      id: Joi.any().strip(),
+      name: Joi.string().max(255).required(),
+      description: Joi.string().max(255),
+      duration: Joi.number().integer().required(),
+      startDate: Joi.date().required(),
+      endDate: Joi.date().min(Joi.ref('startDate')).allow(null),
+      completionDate: Joi.date().min(Joi.ref('startDate')).allow(null),
```
Comment (ngoctay) [inline src/routes/milestones/create.js]: (Review) Milestone completionDate must not be before startDate

File: src/routes/milestones/update.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  },
+  body: {
+    param: Joi.object().keys({
+      id: Joi.any().strip(),
+      name: Joi.string().max(255).required(),
+      description: Joi.string().max(255),
+      duration: Joi.number().integer().required(),
+      startDate: Joi.date().required(),
+      endDate: Joi.date().min(Joi.ref('startDate')).allow(null),
+      completionDate: Joi.date().min(Joi.ref('startDate')).allow(null),
```
Comment (ngoctay) [inline src/routes/milestones/update.js]: (Review) Milestone completionDate must not be before startDate

File: src/routes/milestones/create.js (hunk 3 above)
Comment (coderReview) [reply src/routes/milestones/create.js]: @ngoctay I think it should not be after endDate, wouldn't you agree?

File: src/routes/milestones/update.js (hunk 4 above)
Comment (coderReview) [reply src/routes/milestones/update.js]: @ngoctay I think it should not be after endDate, wouldn't you agree?

Comment (ngoctay) [reply src/routes/milestones/update.js]: I think that it should not be after endDate.
I think startDate and endDate looks like a plan. The completionDate is the actual completed date.
Actual completed date can be before or after then planned endDate.

Comment (coderReview) [reply src/routes/milestones/update.js]: ok

File: src/routes/milestones/update.spec.js [hunk 5]
```diff
@@ -0,0 +1,791 @@
...
+              .then(() => models.Milestone.findById(4))
+              .then((milestone) => {
+                milestone.order.should.be.eql(0);
+
+                done();
+              });
+          }, 3000);
+        });
+    });
+
```
[the remark this answers is not in the document]
Comment (ngoctay) [reply src/routes/milestones/update.spec.js]: I added new tests. Please check the latest commit:

https://github.com/topcoder-platform/tc-project-service/pull/90/commits/157bb75db6516bc5da4af69f1a677ce444e02511

**Pull request merged**